source/common/http/header_map_impl.cc

```diff
@@ -649,5 +655,410 @@ HeaderMapImplUtility::getAllHeaderMapImplInfo() {
+  addViaMove(std::move(ref_key), std::move(new_value));
+  ASSERT(new_value.empty()); // NOLINT(bugprone-use-after-move)
+}
+
+void HeaderMapImpl::setCopy(const LowerCaseString& key, absl::string_view value) {
+  // Replaces the first occurrence of a header if it exists, otherwise adds by copy.
+  // TODO(#9221): converge on and document a policy for coalescing multiple headers.
+  auto* entry = getExisting(key);
+  if (entry) {
+    updateSize(entry->value().size(), value.size());
+    entry->value(value);
+  } else {
+    addCopy(key, value);
+  }
+}
+
+uint64_t HeaderMapImpl::byteSize() const { return cached_byte_size_; }
+
+void HeaderMapImpl::verifyByteSizeInternalForTest() const {
+  // Computes the total byte size by summing the byte size of the keys and values.
+  uint64_t byte_size = 0;
+  for (const auto& map_it : headers_map_) {
+    for (const HeaderEntryImpl& header : map_it.second) {
+      byte_size += header.key().size();
+      byte_size += header.value().size();
+    }
+  }
+  ASSERT(cached_byte_size_ == byte_size);
+}
+
+const HeaderEntry* HeaderMapImpl::get(const LowerCaseString& key) const {
+  // Returns the first occurrence of a Header (if any)
+  const HeaderEagerMapValue* value_list = findMapElement(key);
+  if (value_list == nullptr) {
+    return nullptr;
+  }
+
+  return &value_list->front();
+}
+
+void HeaderMapImpl::mapToSortedVector(const HeaderEagerMap& map,
+                                      std::vector<const HeaderEntryImpl*>& output_vector) const {
+  ASSERT(output_vector.empty());
+  // Add all HeaderEntryImpl elements, and then sort the vector
+  for (const auto& map_entry : map) {
+    for (const HeaderEntryImpl& header_entry : map_entry.second) {
+      output_vector.push_back(&header_entry);
+    }
+  }
+
+  std::sort(output_vector.begin(), output_vector.end(),
+            [](const HeaderEntryImpl* a, const HeaderEntryImpl* b) -> bool {
+              // The iteration order is as follows:
+              // 1. pseudo headers
+              // 2. non-pseudo headers
+              // Within each group, the original header addition order is preserved
+              const bool a_pseudo_header =
+                  !a->key().getStringView().empty() && a->key().getStringView()[0] == ':';
+              const bool b_pseudo_header =
+                  !b->key().getStringView().empty() && b->key().getStringView()[0] == ':';
+              if (a_pseudo_header == b_pseudo_header) {
+                return a->index_ < b->index_;
+              }
+              return a_pseudo_header ? true : false;
+            });
+}
+
+void HeaderMapImpl::iterate(ConstIterateCb cb, void* context) const {
+  // TODO(adisuissa): Because we need to sort the HeaderEntryImpl elements by their index,
+  // we use a temp vector, populate it with the map elements, and sort by the index.
+  // This can be improved by keeping a doubly linked list instead of an index,
+  // and possibly splitting between pseudo-headers and headers.
```

Comment:
What about maintaining a vector in which the vector entry at a given index gives you a reference to the `HeaderEntryImpl` whose index is the vector index? That way, you can easily add new entries (add to the map and push_back), remove (delete the reference at the vector index from the `HeaderEntryImpl`), etc.

Reply:
That could also work. IIUC there will still be "empty" elements in the vector (after removal), but the traversal will be very straightforward.
I'll try it out.

Reply:
I think you would need 2 vectors, one for pseudo headers and one for normal headers. The other option I have considered here is a true linked hash map.

Either of these options will not work with a flat_hash_map though. You will need a node_hash_map for stable elements.

I had thought this implementation would be significantly less code than the O(1) impl, but it's not actually. Interesting stuff and so many trade-offs.

Either way, can you also add more benchmarks around bulk addition like the codecs do? This will be significantly different perf-wise with the different implementations.

Reply:
Yeah, the vector might have missing elements, but I expect it would be relatively dense still, it's not like we tend to add 10k headers and then remove 9990 at random.

There may be a cost to node_hash_map that would be interesting to understand relative to flat_hash_map, or maybe it's a total wash.

Reply:
Thanks for the ideas, agreed on many trade-offs, I can implement the different methods, and we can test it out.
+1 for more benchmarks a la codecs. I'll look into it.

Reply:
One other thought here: we could audit the use of iterate(). In some cases it might be fine to have iteration yield random order. Obviously for serialization in the proxy you'd want to retain it, but IDK if there are other calls to iterate where order doesn't matter.